### PR TITLE
Bugfix: race condition when using multiple sensors

### PR DESCRIPTION
**Pull request by dennisguse**
See https://github.com/OpenTracksApp/OpenTracks/issues/1151#issuecomment-1086693660

Solution: properly synchronize the two methods.

TODO:
- [ ] actually trying it out
- [x] remove TODO commit before merging (catches exception and logs); revert back to crash if time moves backward
- [x] fix tests

### CODE REVIEW

**Comment by dennisguse** [comment]
I am not sure this will solve our race condition.
What we need to make sure is that __all__ calls to `TrackPointCreator.createNow()` result in an in-interruptible call to `TrackPointCreator.onNewTrackPoint()`.
But this is not true for the `GPSHandler` as it first creates a TrackPoint (incl. `createNow()`) and then does some interruptible checks.

In short: TrackPointCreator needs to create the TrackPoints, not the GPSHandler.

**Comment by dennisguse** [comment]
@brianjmurrell ready for testing.
This time the race condition is prevented using an object lock and the whole time generation+storage should never be interupted.


Don't know why this happens and would love to the some documentation. I mean, we have a race condition without threading ;)

**Comment by dennisguse** [comment]
@rgmf Code did not change that much; only thing is that we now the whole adding to TrackPointCreator is synchronized and should therefore enforce an order (and not be interrupted/ multi-threading shouldn't not happen inside).

I am still not sure why this happens exactly:
a) we have multiple threads (IMHO we don't)
b) we have one thread (i.e., managed by one handler), but it might be interrupted to handle another event while it is processing an event.

PS/ Seemingly this is quite a common problem on event-driven architectures.
Here is interesting paper - they did an automated analysis on Google MyTracks and some of the issues sound (still) very familiar. 
We have in the TrackRecordingService some workarounds for this in place (i.e., due to the timer that sends data to the UI).
 https://web.eecs.umich.edu/~nsatish/papers/PLDI-14-AndroidRace.pdf

**Comment by dennisguse** [comment]
@brianjmurrell Can you give this a test drive?
I don't have any sensors at hand....

**Comment by dennisguse** [comment]
Just checked the code and it wasn't working.
Change was more complicated than expected; tests work now locally again.
@brianjmurrell now you could actually test it.

**Comment by rgmf** [comment]
> @rgmf Code did not change that much; only thing is that we now the whole adding to TrackPointCreator is synchronized and should therefore enforce an order (and not be interrupted/ multi-threading shouldn't not happen inside).
> 
> I am still not sure why this happens exactly: a) we have multiple threads (IMHO we don't) b) we have one thread (i.e., managed by one handler), but it might be interrupted to handle another event while it is processing an event.
> 
> PS/ Seemingly this is quite a common problem on event-driven architectures. Here is interesting paper - they did an automated analysis on Google MyTracks and some of the issues sound (still) very familiar. We have in the TrackRecordingService some workarounds for this in place (i.e., due to the timer that sends data to the UI). https://web.eecs.umich.edu/~nsatish/papers/PLDI-14-AndroidRace.pdf

Thank you for the explanation. I'll read the paper ;)

**Comment by dennisguse** [comment]
@rgmf @pstorch I would like to merge this and it should work.
However, I am not 100% certain that we do not run into deadlocks.
No experience there.
So, my proposal is to merge it and create a release to test it in the wild...

File: src/main/java/de/dennisguse/opentracks/services/handlers/TrackPointCreator.java [hunk 1]
```diff
@@ -114,16 +115,28 @@ public void stop() {
         this.context = null;
     }
 
+    public void onChange(@NonNull Location location) {
+        synchronized (this) {
+            TrackPoint trackPoint = new TrackPoint(location, createNow());
+            addSensorData(trackPoint);
+            onNewTrackPoint(trackPoint);
```
Comment (pstorch) [inline src/main/java/de/dennisguse/opentracks/services/handlers/TrackPointCreator.java:122]: First `addSensorData` then `onNewTrackPoint`. In `onNewTrackPoint` line 139 the `addSensorData` is called again. Is that correct?

File: src/main/java/de/dennisguse/opentracks/services/handlers/TrackPointCreator.java [hunk 2]
```diff
@@ -114,16 +115,28 @@ public void stop() {
...
+    }
+
     /**
      * Got a new TrackPoint from Bluetooth only; contains no GPS location.
      */
     @Override
-    public void onChange(SensorDataSet sensorDataSet) {
-        onNewTrackPoint(new TrackPoint(TrackPoint.Type.SENSORPOINT, createNow()));
+    public void onChange(@NonNull SensorDataSet unused) {
+        synchronized (this) {
```
Comment (pstorch) [inline src/main/java/de/dennisguse/opentracks/services/handlers/TrackPointCreator.java:131]: A synchronized block for the whole method? Can this be simplified by adding the synchronized keyword to the method? I think it' equivalent.

**Comment by pstorch** [comment]
> @rgmf @pstorch I would like to merge this and it should work. However, I am not 100% certain that we do not run into deadlocks. No experience there. So, my proposal is to merge it and create a release to test it in the wild...

I don't know if I fully understand the change, but a short test did no harm.

**Comment by rgmf** [comment]
@dennisguse I'm ready to test it ;)

File: src/main/java/de/dennisguse/opentracks/services/handlers/TrackPointCreator.java (hunk 1 above)
Comment (dennisguse) [reply src/main/java/de/dennisguse/opentracks/services/handlers/TrackPointCreator.java:121]: Nice catch; I removed the call - would just be wasting time.

File: src/main/java/de/dennisguse/opentracks/services/handlers/TrackPointCreator.java (hunk 2 above)
Comment (dennisguse) [reply src/main/java/de/dennisguse/opentracks/services/handlers/TrackPointCreator.java:130]: Good question; not sure.
Just checked the documentation and making the onChange()-methods would have the same effect.
Was not aware of this - thanks for the hint :D

https://docs.oracle.com/javase/tutorial/essential/concurrency/syncmeth.html

Comment (dennisguse) [reply src/main/java/de/dennisguse/opentracks/services/handlers/TrackPointCreator.java]: Just added a commit changing this :D

**Pull request merged**